### PR TITLE
Improved undo

### PR DESCRIPTION
**Pull request by openroadie**

### CODE REVIEW

**Comment by openroadie** [comment]
The new code is not yet enabled since I need to add better statistics (it will also enable us to correctly count number of moves made vs. number of moves that were rejected/undone) but I would like this PR merged so it does not become an unmanageable mess. The next steps are:

- Add the missing corner cases/more thorough testing
- Once ORFS and maybe even secure ORFS successfully passes, we can move to the new undo method
- Delete the existing undo code if there are no issues

File: src/rsz/test/repair_fanout7.ok [hunk 1]
```diff
@@ -12,9 +12,9 @@ load DFF_X1 CK D 300000 metal3 1000
...
-[INFO RSZ-0045] Inserted 78 buffers, 1 to split loads.
-[INFO RSZ-0041] Resized 527 instances.
-[INFO RSZ-0049] Cloned 7 instances.
+[INFO RSZ-0045] Inserted 41 buffers, 1 to split loads.
+[INFO RSZ-0041] Resized 312 instances.
+[INFO RSZ-0049] Cloned 4 instances.
 [WARNING RSZ-0062] Unable to repair all setup violations.
 Repair timing output passed/skipped equivalence test
-worst slack -3.75
+worst slack -2.84
```
Comment (maliberty) [inline src/rsz/test/repair_fanout7.ok:20]: If "The new code is not yet enabled" why are there changes to the results?

File: src/rsz/include/rsz/Resizer.hh [hunk 2]
```diff
@@ -514,13 +514,16 @@ protected:
...
                                 SteinerPt pt,
                                 const ParasiticAnalysisPt *parasitics_ap);
-
-  bool replaceCell(Instance *inst,
-                   LibertyCell *cell,
-                   bool journal);
-
   void findResizeSlacks1();
+  
+public:
```
Comment (maliberty) [inline src/rsz/include/rsz/Resizer.hh:519]: Consolidate the public & protected sections

File: src/rsz/src/Resizer.cc [hunk 3]
```diff
@@ -2701,38 +2718,98 @@ Resizer::journalSwapPins(Instance *inst, LibertyPort *port1,
...
+    if (network_->direction(pin)->isInput()) {
+      Instance *inst = network_->instance(pin);
+      auto term_port = network_->port(pin);
+      sta_->connectPin(inst, term_port, original_out_net);
+    }
+  }
+  // Final cleanup
+  // sta_->deleteNet(clone_out_net);
+  sta_->deleteInstance(cloned_inst);
+  sta_->graphDelayCalc()->delaysInvalid();
```
Comment (maliberty) [inline src/rsz/src/Resizer.cc:2795]: Shouldn't sta do this automatically?

File: src/rsz/src/Journal.cc [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+UndoBuffer::UndoBuffer(Instance* inst)
+{
+  // Original cell used for resize and the instance
+  buffer_inst_ = inst;
+}
+
+int UndoBuffer::UndoOperation(Resizer *resizer)
+{
+  // Remove the inserted buffer
+  resizer->removeBuffer(const_cast<Instance*>(buffer_inst_));
```
Comment (maliberty) [inline src/rsz/src/Journal.cc:73]: Why is a const_cast needed as the field isn't const?

File: src/rsz/src/Journal.cc [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    resizer_ = resizer;
+    logger_ = logger;
+}
+
+void Journal::begin()
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal begin");
+  while (!journal_stack_.empty()) {
+    journal_stack_.pop();
+  }
```
Comment (maliberty) [inline src/rsz/src/Journal.cc:148]: Why not "journal_stack_ = {};" like Jounal::end() ?

File: src/rsz/src/Journal.cc [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  // TODO We need to do something with cloned instances here
+  journal_stack_ = {};
+}
+
+void Journal::swapPins(Instance* inst, LibertyPort* port1,
+                              LibertyPort* port2)
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal swap pins {} ({}->{})",
+             network()->pathName(inst), port1->name(), port2->name());
+  std::unique_ptr<Undo> element(new UndoPinSwap(inst, port1, port2));
```
Comment (maliberty) [inline src/rsz/src/Journal.cc:163]: Use std::make_unique

File: src/rsz/src/Resizer.cc (hunk 3 above)
Comment (openroadie) [reply src/rsz/src/Resizer.cc:2795]: I am assuming you mean the delaysInvalid call. Fixed.

File: src/rsz/test/repair_fanout7.ok (hunk 1 above)
Comment (openroadie) [reply src/rsz/test/repair_fanout7.ok:20]: General cleanup and restructuring of the code so we can actually switch over at some point.

**Comment by openroadie** [comment]
@maliberty This is finally ready. The next step would be to do some testing of the new undo mode and then migrate to it (and delete all the existing undo code). The big difference between new/old Undo is that the new code actually does use a stack (instead of the old ad hoc approach where certain undo operations can simply not be supported).

File: src/rsz/test/repair_fanout7.tcl [hunk 7]
```diff
@@ -34,5 +34,4 @@ estimate_parasitics -placement
 report_worst_slack -max
 write_verilog_for_eqy repair_fanout7 before "None"
 repair_timing -setup -repair_tns 100 -skip_pin_swap
-run_equivalence_test repair_fanout7 ./Nangate45/work_around_yosys/ "None"
```
Comment (maliberty) [inline src/rsz/test/repair_fanout7.tcl:37]: Why remove this?

File: src/rsz/test/repair_fanout7.ok [hunk 8]
```diff
@@ -12,9 +12,8 @@ load DFF_X1 CK D 300000 metal3 1000
...
-[INFO RSZ-0045] Inserted 78 buffers, 1 to split loads.
-[INFO RSZ-0041] Resized 527 instances.
-[INFO RSZ-0049] Cloned 7 instances.
+[INFO RSZ-0045] Inserted 40 buffers, 1 to split loads.
+[INFO RSZ-0041] Resized 364 instances.
+[INFO RSZ-0049] Cloned 4 instances.
 [WARNING RSZ-0062] Unable to repair all setup violations.
-Repair timing output passed/skipped equivalence test
-worst slack -3.75
+worst slack -2.84
```
Comment (maliberty) [inline src/rsz/test/repair_fanout7.ok:19]: I thought this is off by default - why is there a diff?

File: src/rsz/src/Journal.hh [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+enum class BufferedNetType;
+using BufferedNetPtr = std::shared_ptr<BufferedNet> ;
+using BufferedNetSeq = vector<BufferedNetPtr>;
+
+class Undo
+{
+ public:
+  // base class for undo elements
+  Undo() = default;
+  virtual int UndoOperation(Resizer *resizer) = 0;
```
Comment (maliberty) [inline src/rsz/src/Journal.hh:71]: This needs a virtual dtor

Comment (maliberty) [inline src/rsz/src/Journal.hh:71]: What does the return value indicate?

File: src/rsz/src/Journal.cc [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    logger_->info(RSZ, 97, "Cloned {} instances.", cloned_gate_count_);
+  }
+  }
+
+sta::Network * Journal::network()
+{
+  return resizer_->network_;
+}
+
+dbSta *Journal::sta()
```
Comment (maliberty) [inline src/rsz/src/Journal.cc:232]: What is this used for?

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/rsz/test/repair_fanout7.tcl (hunk 7 above)
Comment (openroadie) [reply src/rsz/test/repair_fanout7.tcl:37]: Because it fails on CENTOS7 and I do not know whether it is related to an eqy issue or some other problem.

File: src/rsz/test/repair_fanout7.ok (hunk 8 above)
Comment (openroadie) [reply src/rsz/test/repair_fanout7.ok:19]: What is "off my default"

File: src/rsz/src/Journal.cc (hunk 10 above)
Comment (openroadie) [reply src/rsz/src/Journal.cc:232]: Removed for now. Will pit it back when needed.

File: src/rsz/src/Journal.hh (hunk 9 above)
Comment (openroadie) [reply src/rsz/src/Journal.hh:71]: Not much for the moment but we might need success/failure once we try to make this the default.

File: src/rsz/test/repair_fanout7.ok (hunk 8 above)
Comment (openroadie) [reply src/rsz/test/repair_fanout7.ok:19]: The diff comes from some cleanup that happened during the implementation of the infrastructure to redo the "undo" operation. The results are better.

Comment (maliberty) [reply src/rsz/test/repair_fanout7.ok:19]: typo : off by default

Comment (maliberty) [reply src/rsz/test/repair_fanout7.ok:19]: Does "The results are better." imply you ran a secure CI?

Comment (openroadie) [reply src/rsz/test/repair_fanout7.ok:19]: No. All it implies is that these particular results are better (-3.75 vs. -2.84). If you think there is a need to run the secure CI for this change then I can try to remember how to do that. LMK.

**Comment by maliberty** [comment]
I went ahead and started a secure CI

**Comment by openroadie** [comment]
Thank you @maliberty . Much appreciated.

**Comment by maliberty** [comment]
aes/ng45:  Error: cts.tcl, 111 ERROR: A problem occurred during equivalence check.
aes_lvt/asap7: [ERROR] finish__timing__wns_percent_delay fail test: -12.204497 >= -10.0

The first needs investigating but the second could be updated.

**Comment by openroadie** [comment]
> aes/ng45: Error: cts.tcl, 111 ERROR: A problem occurred during equivalence check. aes_lvt/asap7: [ERROR] finish__timing__wns_percent_delay fail test: -12.204497 >= -10.0
> 
> The first needs investigating but the second could be updated.

Thank you @maliberty . I agree with you and luckily can run aes/n45 locally. I will investigate and report back. Sorry for the delay in replying (been kinda sick this week).

**Comment by maliberty** [comment]
stale - replaced by odb undo
